core/src/main/java/xyz/gianlu/librespot/player/Player.java

```diff
@@ -554,20 +558,12 @@ public void trackChanged(@NotNull String playbackId, @Nullable TrackOrEpisode me
         events.trackChanged();
 
         session.eventService().newPlaybackId(state, playbackId);
-        metrics = new PlaybackMetrics(state.getCurrentPlayableOrThrow(), playbackId, state);
-        metrics.startedHow(startedReason, state.getPlayOrigin().getFeatureIdentifier());
-        metrics.startInterval(pos);
+        startMetrics(playbackId, startedReason, pos);
     }
 
     @Override
-    public void trackPlayed(@NotNull PlaybackMetrics.Reason endReason, @NotNull PlayerMetrics playerMetrics, int willEndAt) {
-        if (metrics != null) {
-            metrics.endedHow(endReason, state.getPlayOrigin().getFeatureIdentifier());
-            metrics.endInterval(willEndAt);
-            metrics.update(playerMetrics);
-            session.eventService().trackPlayed(metrics);
-            metrics = null;
-        }
+    public void trackPlayed(@NotNull String playbackId, @NotNull PlaybackMetrics.Reason endReason, @NotNull PlayerMetrics playerMetrics, int when) {
+        endMetrics(playbackId, endReason, playerMetrics, when);
```

Comment:
endMetrics called in trackPlayed

Reply:
That's not the same method, they just have the same name.

Reply:
yeah my bad, sorry :)